### PR TITLE
Handle merge'ing empty diffs specially

### PR DESCRIPTION
**Pull request by schwers-zz**
This changes fixes how empty state diffs are treated in merge. In strict mode, empty state diffs should not be creating new objects because state did not change. In replace mode, empty state diffs should be completely overwriting state.

Relatedly, we could also consider using https://facebook.github.io/react/docs/update.html

👓 @nramadas

### CODE REVIEW

File: lib/modules/merge.js [hunk 1]
```diff
@@ -2,13 +2,24 @@ import { isPlainObject } from 'lodash/lang';
...
 
+  // Empty state diffs get special handling. They shouldn't cause a new state
+  // object to be created in strict mode, but in replace mode we should return
+  // the new empty object.
+  if (!Object.keys(stateDiff).length) {
+    if (emptyDict === 'replace') {
+      return stateDiff;
+    }
+
+    return state;
```
Comment (nramadas) [inline lib/modules/merge.js]: this logic isn't exactly correct. emptyDict has three values, 'skip', 'replace', and 'strict'. this should read:

``` es6
  if (!Object.keys(stateDiff).length) {
    if (emptyDict === 'replace') {
      return stateDiff;
    } else if (emptyDict === 'skip') 
      return state;
    }
  }
```

**Comment by nramadas** [comment]
:haircut:

File: lib/modules/merge.js (hunk 1 above)
Comment (schwers-zz) [inline lib/modules/merge.js]: Can you elaborate on the difference between `skip` and `strict`? It seems like `skip` should be returning the old state, which makes sense. But should't `strict` be doing so as well? If both and `strict` and `skip` return state what I have seems fine

Comment (nramadas) [inline lib/modules/merge.js]: `strict` will merge in the new dict, causing the reference to change. it does the merge no matter what, hence the 'strict'. it's the default behavior; objects are merged, empty or not.

**Comment by schwers-zz** [comment]
💇

**Pull request merged**